### ISSUE TITLE
Domain manager can reset admin password

### ISSUE THREAD

**Issue by Willow0349** [Mailu/Mailu#2685]
### Environment

- [x] docker-compose
- [ ] kubernetes
- [ ] docker swarm

### Version

- Version: `1.9`

## Description
Domain manager can change password on admin users

## Replication Steps
create a admin user on domain.tld
create a  domain manager on domain.tld
login as domain manager on domain.tld
try to change admin users password as domain manager

## Observed behaviour
admin password is reset and i can login with the new password

## Expected behaviour
domain manager is not allowed to change admin password because admin has higher priveleges than domain manager

**Comment by fastlorenzo** [comment]
Another scenario that poses an issue:

user1@domain1 is domain manager of domain1
user2@domain1 is domain manager of domain2

user1@domain1 can reset the password of user2@domain1 and take control of domain2

A possible solution would be to deny a password reset as domain manager if:
1. The target user for the password reset is an admin user
2. The target user for the password reset is domain manager of other domains (except if the target user is the user doing the password reset, ofc)

**Comment by nextgens** [comment]
@Willow0349 Thank you for your report.

We are currently considering how to best address this. Other functions (such as deletion and token creation) are also potentially problematic.

The easy workaround is to ensure that all your global admins are on a separate (potentially non-existing) domain; we could document this.

**Comment by nextgens** [comment]
This has now been discussed on #mailu-dev : The current view is that we should document the behavior and warn (by asking confirmation) when rights are about to be granted to a user.

Fixing it any other way would involve a substantial amount of code that we would be uncomfortable backporting.

**Comment by elandorr** [comment]
Thank you for reporting this big issue! Goes to show messing around is important before relying on security.

To make maintenance easier and not just have one guy do all the work, delegations are important, but they must not under any circumstance override expected permissions.

**Comment by Jayd603** [comment]
So... is this still an issue?

**Comment by maximilize** [comment]
Still reproduces at HEAD. `domain_admin` (`core/admin/mailu/ui/access.py:59-62`) authorizes purely on the *target's* domain — `domain in current_user.get_managed_domains()` — with no rank check. So `user_edit` (`ui/views/users.py:56`) and `user_password` (`:142`), both `@access.domain_admin(models.User, 'user_email')`, let a domain manager edit/reset a **global admin** (or a manager of another domain) who merely holds a mailbox in the managed domain → takeover, exactly as @fastlorenzo described.

The enforcement route is small, not "a substantial amount of code": @fastlorenzo's rule fits inside `domain_admin` — when the target is a `User` other than the actor, additionally require `not target.global_admin` and `set(target.get_managed_domains()) <= set(current_user.get_managed_domains())`. The `manager_of`/`global_admin` data is already there (`models.py:567`, `:703`).

If you prefer enforcement over the documented-workaround route, I'll send a PR with that rule plus a `tests/anonmail` regression test (domain manager → reset a global admin's password → assert 403).

**Comment by nextgens** [comment]
You're right to point out that 'substantial' is the wrong adjective here, I should probably have used 'complex' instead.

Let me point out where the complexity arises: maybe we want to allow some of the operations (to be defined, as written above it's not just the change of password that needs to be gated) despite the lack of authorisation. Take the example an account that is the admin of a domain that we know to be compromised: should other privileged users that are not domain admin be able to disable it? We probably do not want to allow password or token changes (to avoid the privesc this ticket is about) but maybe we ought to let 'set the disabled bit' through... or maybe we don't (because DoS) and we should keep it simple.

Please do send a PR and we can discuss it with code.
